imagenet/main.py

```diff
@@ -76,33 +76,33 @@ def __call__(self, path: Union[str, pathlib.Path]) -> Dict[str, Any]:
         return dict(label=label, wnid=wnid, cls=cls)
 
 
-class ImageNet:
-    def __init__(
-        self,
-        root: Union[str, pathlib.Path],
-        *,
-        split: str = "train",
-        decoder: Optional[str] = "pil",
-    ):
-        self.root = pathlib.Path(root)
-        self.split = split
-        self._meta = _ImageNetMeta(self.root, split=self.split)
+def ImageNet(root: Union[str, pathlib.Path],
```

Comment:
Not sure if this is possible in general. For most datasets we have additional logic that I simply skipped in this repo such as the donwload functionality. I've discussed this with @fmassa and we agreed that a class makes it most readable. It also allows us to impose some structure.

@ejguan Could you elaborate why you think this is warranted?

Reply:
Implementing several DataPipe to support the functionality like downloader or reader is also in our plan. By attaching this kind of DataPipe in front of the pipeline does support the goal.

There are several reasons for using a function to return a stacked DataPipe rather than a class:
- IMO, it will be easier for new DataLoader to manipulate the data pipeline using this stacked DataPipe. It will be useful as DataLoader can take in charge of sharding, joining, and etc.
- Even if we let class `ImageNet` inheriting `IterDataPipe`, the problem is all the prior `DataPipe`s are hidden behind this class instance. We can not extract the sequence of `DataPipe` from it. 
- Based on the idea of data pipeline, the construction of each `DataPipe` should take at least one prior `DataPipe` as an argument.
- This structure doesn't prevent you to create the Class (but inheriting `IterDataPipe`) and impose some structure. You can create a DataPipe attached at the end of pipeline, rather than putting all prior `DataPipe`s within the DataPipe

Based on your opinion, I propose an alternative way to create pipeline. How about this structure below?
```py
class ImageNetDataPipe(IterDataPipe):
    def __init__(self, datapipe, root, split):
        self.datapipe = datapipe
        self.root = pathlib.Path(root)
        self.split = split
        self.meta_func = _ImageNetMeta(self.root, split=self.split)
        ...
    def __iter__(self):
        for path, image in self.datapipe:
            sample = dict(image_path=path, image=image)
            if self.meta_func is not None:
                sample.update(self.meta_func(path))
            yield sample

def ImageNet(root, split, decoder):
    datapipe = dp.iter.LoadFilesFromDisk((path))
    ...
    if decoder is not None:
        datapipe = dp.iter.RoutedDecoder(datapipe, handlers=[imagehandler(decoder)])
    datapipe = ImageNetDataPipe(datapipe, root, split)
    return datapipe
```

BTW, the reason I didn't add this `ImageNet` DataPipe was all the functionality could be replaced by the existing `DataPipe` in main repo.

Reply:
> Implementing several DataPipe to support the functionality like downloader or reader is also in our plan. By attaching this kind of DataPipe in front of the pipeline does support the goal.

Without further research I think having downloader datapipes would indeed eliminate the need for a custom dataset class.

Reply:
Thanks for the feedback.
Please correct me if I am wrong. In torchvision, there are two sources for downloading, which needs to be implemented as separate or integrated `DownloaderDataPipe`:
- google drive
- normal url
Besides, we could use `ReaderDataPipe` to get file stream from NFS. Does torchvision have some Dataset in mind requiring the `Reader` feature?

Reply:
> Please correct me if I am wrong. In torchvision, there are two sources for downloading, which needs to be implemented as separate or integrated `DownloaderDataPipe`:
> 
>  - google drive
>  - normal url

Correct. We also host one dataset which is simultaneously hosted on Baidu Cloud, but IMO that is not important enough to support it yet. 

>  Besides, we could use `ReaderDataPipe` to get file stream from NFS. Does torchvision have some Dataset in mind requiring the `Reader` feature?

Right now we simply point the dataset to a root directory and read from there. Do we need special treatment if the dataset is on NFS?

Reply:
Plus, it would be awesome if we had parallel downloads. There it would be again useful to be able to first download the meta data and the actual data last.

Reply:
> Do we need special treatment if the dataset is on NFS?

No, and just want to confirm.

> Plus, it would be awesome if we had parallel downloads.

This one should definitely be achieved as long as the data source is supported.

> There it would be again useful to be able to first download the meta data and the actual data last.

Good catch. But, it depends on whether the specific dataset on cloud/web provides this API to separate meta and actual data, so maybe it would not fall into main repo for general purpose.
And this feature can be achieved by users to combine two downloader datapipes (attaching one meta data pipeline within the actual data pipeline).